### PR TITLE
feat: reads shiny item stats to optional

### PR DESCRIPTION
**Pull request by geekazodium**
there's a 99% chance the commit title is not using the correct prefix, pls correct me if I'm titling this wrong.

### CODE REVIEW

**Comment by kristofbolyai** [comment]
@magicus should review this.

This is good, but we can't use this for chat items nor to fix item handler atm.

**Comment by geekazodium** [comment]
> @magicus should review this.
> 
> This is good, but we can't use this for chat items nor to fix item handler atm.

I'm going to check the issues for this, hopefully I can help, if you could pls reference the issue with info about the item handler in issue tracker so I don't do something dumb like read the wrong issue and get massively confused about what's going on.

**Comment by kristofbolyai** [comment]
https://github.com/Wynntils/Artemis/issues/1934 is the issue, but it is not an easy fix. If it was easy, we wouldn't have made an issue, but fixed it.

**Comment by geekazodium** [comment]
alright I'm going to clothesline into the wall I just realized a detail I missed with the itemhander issue.

**Comment by geekazodium** [comment]
> #1934 is the issue, but it is not an easy fix. If it was easy, we wouldn't have made an issue, but fixed it.

part 1/2 complete, mod should now ignore when checking if items are the same.

**Comment by geekazodium** [comment]
P.S. the way I obtained the line where the shiny mythic statistic is might need 1. testing, 2. an entire rewrite because of issues when writing that code (should be in comments)

**Comment by geekazodium** [comment]
also I think the spotless check is failing but I can't find where in the code is causing it.

**Comment by geekazodium** [comment]
Nevermind I fixed the * import issue.

File: common/src/main/java/com/wynntils/models/wynnitem/parsing/WynnItemParser.java
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/models/wynnitem/parsing/WynnItemParser.java]: What's the point of having the no capture pattern?

File: common/src/main/java/com/wynntils/utils/mc/LoreUtils.java
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/utils/mc/LoreUtils.java]: We disallow static imports, please remove them from the code.

**Review by kristofbolyai — COMMENTED** [review]
Please run `spotlessCheck` and `spotlessApply`, your code needs some formatting.

File: common/src/main/java/com/wynntils/utils/mc/LoreUtils.java
Comment (geekazodium) [reply common/src/main/java/com/wynntils/utils/mc/LoreUtils.java]: okay, got it

File: common/src/main/java/com/wynntils/models/wynnitem/parsing/WynnItemParser.java
Comment (geekazodium) [reply common/src/main/java/com/wynntils/models/wynnitem/parsing/WynnItemParser.java]: there are cases where the capture return values are discarded, should I still use the capture pattern anyways? I am new to using regex.

File: common/src/main/java/com/wynntils/handlers/item/ItemHandler.java
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/handlers/item/ItemHandler.java]: Can you hold off on adding the event until it is actually used?

File: common/src/main/java/com/wynntils/handlers/item/ItemHandler.java
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/handlers/item/ItemHandler.java]: You can just use pairs, although if you remove the event, this is not needed atm.

File: common/src/main/java/com/wynntils/models/wynnitem/parsing/WynnItemParser.java
Comment (kristofbolyai) [inline common/src/main/java/com/wynntils/models/wynnitem/parsing/WynnItemParser.java]: debug

Comment (geekazodium) [reply common/src/main/java/com/wynntils/models/wynnitem/parsing/WynnItemParser.java]: *why did I forget to remove this.

File: common/src/main/java/com/wynntils/models/wynnitem/parsing/WynnItemParser.java
Comment (kristofbolyai) [reply common/src/main/java/com/wynntils/models/wynnitem/parsing/WynnItemParser.java]: capture groups are only relevant if you actually get them. you can just ignore them existing, if you dont need them.

Comment (geekazodium) [reply common/src/main/java/com/wynntils/models/wynnitem/parsing/WynnItemParser.java]: okay, thx for letting me know.

File: common/src/main/java/com/wynntils/handlers/item/ItemHandler.java
Comment (geekazodium) [reply common/src/main/java/com/wynntils/handlers/item/ItemHandler.java]: ok :thumbs_up:

**Comment by geekazodium** [comment]
I think that's all of the immediate issues fixed, should probably do a bit of more testing on shiny mythics to verify this currently works, it's 1 AM right now though so I must sleep.

**Comment by kristofbolyai** [comment]
@P0keDev What do you think of this? Do you want to implement it in a different way?

File: common/src/main/java/com/wynntils/models/wynnitem/parsing/WynnItemParser.java
Comment (magicus) [inline common/src/main/java/com/wynntils/models/wynnitem/parsing/WynnItemParser.java]: What does the `?` do after the `*`? Also, I'd suggest using `+` instead of `*`; we'd want at least one character here.

File: common/src/main/java/com/wynntils/models/gear/GearChatEncoding.java
Comment (magicus) [inline common/src/main/java/com/wynntils/models/gear/GearChatEncoding.java:204]: That is never going to happen. We cannot change the "chat transfer" protocol to support new functions, due to the lack of version handling in the protocol. :-(

Comment (magicus) [reply common/src/main/java/com/wynntils/models/gear/GearChatEncoding.java:204]: Or rather, it basically requires us to design a new protocol. And figure out how to handle old and incompatible clients.

**Review by magicus — APPROVED** [review]
I think this is basically looking good. One open question is if we really should store shinyStat with the GearInstance. I was thinking that the other data there are unchanging, but it is not strictly true I realize now -- rerolls can cause changes. So I guess we'll accept that shiny stats can change as well.

File: common/src/main/java/com/wynntils/models/wynnitem/parsing/WynnItemParser.java
Comment (geekazodium) [reply common/src/main/java/com/wynntils/models/wynnitem/parsing/WynnItemParser.java]: okay the unneeded char in the regex has been removed 👍

File: common/src/main/java/com/wynntils/models/gear/GearChatEncoding.java
Comment (geekazodium) [reply common/src/main/java/com/wynntils/models/gear/GearChatEncoding.java:204]: I mean my first instinct is to suggest attaching the shiny stat to the the end of the existing item string with a spacebar in-between so older clients will just parse the item the same but updated clients can see the stat by checking if there's more unicode encoded item info after the space char.

(but would that be a janky solution idk.)

Comment (magicus) [reply common/src/main/java/com/wynntils/models/gear/GearChatEncoding.java:204]: The real solution is that we need to create a new "protocol" that has some kind of version information, or is extendable in some other way, so we can continue adapting for future Wynn changes. But that is a quest for another day...

Comment (geekazodium) [reply common/src/main/java/com/wynntils/models/gear/GearChatEncoding.java:204]: following through that idea, maybe encoding the statistic name with a single byte should be fine I doubt I'll see the day in which wynncraft has 255 or something statistics on shiny mythics, but who knows, if we suppose that wynncraft will eventually have the 128th unique shiny mythic statistic, we can reserve the first bit in that as a flag value just in case this happens and if that bit is true we can check the next byte as well or something. (can't wait for the day when wynncraft releases the 128th unique shiny mythic stat, and we don't have to worry about backwards compatibility in that very specific scenario.

(but haven't I made a janky enough of an idea maybe shouldn't make it jankier.

Comment (magicus) [reply common/src/main/java/com/wynntils/models/gear/GearChatEncoding.java:204]: Please just remove this comment, and then you're good to go.

Comment (geekazodium) [reply common/src/main/java/com/wynntils/models/gear/GearChatEncoding.java:204]: tldr: <itemname> + ' ' + new stuff here

Comment (magicus) [reply common/src/main/java/com/wynntils/models/gear/GearChatEncoding.java:204]: Problem is, there are already other information we can't properly pass using this chat protocol. So we need to fix that as well.

Comment (geekazodium) [reply common/src/main/java/com/wynntils/models/gear/GearChatEncoding.java:204]: oh, that makes sense.

**Comment by geekazodium** [comment]
update: this may not be working as intended because it appears the this shiny divzer's mobs killed stats are not being read properly.

**Comment by geekazodium** [comment]
*correction, it works but it doesn't update it when a mob is killed.

**Comment by geekazodium** [comment]
I think I know the issue... I was planning to make it so that the shiny mythic stat update event will trigger that stat to be parsed, and just... forgot.

**Comment by geekazodium** [comment]
can't merge until that issue has been fixed

**Comment by geekazodium** [comment]
> I think this is basically looking good. One open question is if we really should store shinyStat with the GearInstance. I was thinking that the other data there are unchanging, but it is not strictly true I realize now -- rerolls can cause changes. So I guess we'll accept that shiny stats can change as well.

... so turns out you can't modify values on records* so...

* -> I can always just make a something similiar to an optional but non final to work around that, or should I be storing this somewhere else?

**Comment by geekazodium** [comment]
*doing a bit of testing and it works with my shiny divzer I bought from trade market for 2stx

**Comment by geekazodium** [comment]
> *doing a bit of testing and it works with my shiny divzer I bought from trade market for 2stx

tldr: not worth but I have it sooooo............

**Comment by geekazodium** [comment]
I don't know if my implementation might be messy tho..

**Review by P0keDev — CHANGES_REQUESTED** [review]
This approach in fixing `ItemHandler` is the wrong one, IMO. See #1957

**Comment by geekazodium** [comment]
> This approach in fixing `ItemHandler` is the wrong one, IMO. See #1957

I see, I was suggesting to have the shiny stat saved to an accessible value so it can also be sent and read with chat encoding #1676, although I can see that yeah your fix with #1957 is probably much better at preventing shiny statistics from messing up the `ItemHandler` (If I'm understanding this correctly).

**Comment by magicus** [comment]
The most critical problem with shiny stats is that it breaks the ItemHandler. This is addressed in #1977 (as an alternative to #1957). Once that is resolved , I can start looking at how to integrate the extra value shiny stats give into the Wynntils models. This is a start in that direction, but we need to do more. I'll come back with a more precise review of this PR when we are at that point.

**Comment by magicus** [comment]
This will be tricky to get right, both in terms of adapting this PR to the changes made in the Itemhandler recently, and to getting ShinyStats to work with the system we have. I have picked out the parts from this PR that I think is a relevant contribution, and will continue massaging them into something we can merge in #2006. I will to make sure you get proper credit for your part of that work, @geekazodium, and thanks for your contribution!

**Comment by kristofbolyai** [comment]
Supersed by #2006

**Comment by kristofbolyai** [comment]
Thank you for contributing, your code will be partially merged by that PR.

**Pull request closed without merging**
